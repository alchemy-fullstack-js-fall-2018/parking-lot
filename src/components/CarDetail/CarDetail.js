import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './CarDetail.css';

export default class CarDetail extends Component {

  static propTypes = {
    car: PropTypes.object.isRequired,
    carArrives: PropTypes.func.isRequired,
    carDeparts: PropTypes.func.isRequired
  };

  render() {
    const { plate, color, make, model, lotUseCount, isInLot } = this.props.car;
    const { carArrives, carDeparts } = this.props;
    const status = isInLot ? 'Parked' : 'Absent';

    return (
      <div>
        <h2>Car Detail</h2>
        <p><span className={styles.bold}>Plate:</span> {plate}</p>
        <p><span className={styles.bold}>Description:</span> {color} {make} {model}</p>
        <p><span className={styles.bold}>Times in lot:</span> {lotUseCount}</p>
        <p><span className={styles.bold}>Status:</span> {status}</p>
        {isInLot && <button onClick={() => carDeparts(plate)}>Car has departed</button>}
        {!isInLot && <button onClick={() => carArrives(plate)}>Car has arrived</button>}
      </div>
    );
  }

}
