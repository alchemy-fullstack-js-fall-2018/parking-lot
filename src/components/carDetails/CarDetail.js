import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { ROUTES } from '../../routes';

const carDetail = ({ car, inLot, lotCarArrived, lotCarDeparted, lotUsage }) => {

  const { make, model, plate } = car;

  const handleClick = event => {
    event.preventDefault();
    if(!inLot) lotCarArrived(car)
    else lotCarDeparted(plate)
  }

  console.log(inLot);
  const status = inLot ? 'Yes' : 'No';
  const addRemoveButton = inLot ? 'Remove' : 'Add';

  return(
    <Fragment>
      <table>
        <thead>
          <tr>
            <th>Make</th>
            <th>Model</th>
            <th>Plate</th>
            <th>In Lot:</th>
            <th>Lot Usage:</th>
            <th>Add or remove from lot</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th>{make}</th>
            <th>{model}</th>
            <th>{plate}</th>
            <th>{status}</th>
            <th>{lotUsage}</th>
            <th>
              <button onClick={handleClick}>{addRemoveButton}</button>
            </th>
          </tr>
        </tbody>
      </table>
    </Fragment>
  );
}

carDetail.propTypes = {
  car: PropTypes.object.isRequired,
  inLot: PropTypes.bool.isRequired,
  lotCarArrived: PropTypes.func.isRequired,
  lotCarDeparted: PropTypes.func.isRequired,
  lotUsage: PropTypes.number.isRequired
}

export default carDetail;
