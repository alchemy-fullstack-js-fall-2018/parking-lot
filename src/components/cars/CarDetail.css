.header {
  text-align: center;
}

.outer {
  display: flex;
  align-items: center;
}

.wrapper {
  display: block;
  text-align: left;
  margin-left: 4.5em;
}

.inner {
  width: 30%;
  display: inline-block;
  margin: 0 auto;
  text-align: center;
  border: #3a4145 solid 2px;
}

.button{
	color: black;
  background: none;
  border-radius: 0;
  padding: 1.2em 5em;
  font-size: 12px;
  transition: background-color 0.3s, box-shadow 0.3s, color 0.3s;
  margin: 1em;
  box-shadow: inset 0 0 1em rgba(58,65,69,0.5), 0 0 1em rgba(58,65,69,0.5);
  border: #3a4145 solid 2px;

    &:hover {
        color: white;
        background-color: #3a4145;
        box-shadow: inset 0 0 0 rgba(58,65,69,0.5), 0 0 1.5em rgba(58,65,69,0.5);
    }

}