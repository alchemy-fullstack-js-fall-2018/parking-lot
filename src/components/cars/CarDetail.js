import React, { Fragment } from 'react';
import { Redirect } from 'react-router-dom';
import { ROUTES } from '../../routes/routes';
import PropTypes from 'prop-types';

function CarDetail({ car, inLot, addToLot, removeFromLot, lotUsage }) {
  if(!car) return <Redirect to={ROUTES.HOME.linkTo()} />;

  const { make, model, plate } = car;
  const lotStatus = inLot ? 'Yes' : 'No';
  const buttonText = inLot ? 'Remove from Lot' : 'Add to Lot';

  const handleClick = event => {
    event.preventDefault();
    if(!inLot) {
      addToLot(make, model, plate);
    } else {
      removeFromLot(plate);
    }
  };


  return (
    <Fragment>
      <table width="100%">
        <tbody>
          <tr>
            <th>Make</th>
            <th>Model</th>
            <th>Plate</th>
            <th>In Lot Now</th>
            <th>Times Used</th>
          </tr>
          <tr>
            <th>{make}</th>
            <th>{model}</th>
            <th>{plate}</th>
            <th>{lotStatus}</th>
            <th>{lotUsage}</th>
          </tr>
        </tbody>
      </table>
      <button onClick={handleClick}>{buttonText}</button>
    </Fragment>
  );
}

CarDetail.propTypes = {
  car: PropTypes.object.isRequired,
  inLot: PropTypes.bool.isRequired,
  addToLot: PropTypes.func.isRequired,
  removeFromLot: PropTypes.func.isRequired,
  lotUsage: PropTypes.number.isRequired
};

export default CarDetail;
