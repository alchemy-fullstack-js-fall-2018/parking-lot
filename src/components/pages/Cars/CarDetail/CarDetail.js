import React from 'react';
import PropTypes from 'prop-types';

const CarDetail = ({ car }) => {
  const { plate, make, model, inLot, lotUses } = car;
  return (
    <div>
      <p>plate: {plate}</p>
      <p>make: {make}</p>
      <p>model: {model}</p>
      <p>currently in lot: {inLot}</p>
      <p>times in lot: {lotUses}</p>
    </div>
  );
};

CarDetail.propTypes = {
  car: PropTypes.object.isRequired
};

export default CarDetail;
