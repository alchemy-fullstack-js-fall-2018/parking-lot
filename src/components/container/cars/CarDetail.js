import { connect } from 'react-redux';
import CarDetail from '../../presentational/cars/CarDetail.jsx';
import { getCar } from '../../../data/state/selectors/cars';
import { carArrived, carDeparted } from '../../../data/state/actions/cars';

const mapStateToProps = (state, props) => ({
  car: getCar(state, props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  carDeparted: id => dispatch(carDeparted(id)),
  carArrived: id => dispatch(carArrived(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarDetail);
