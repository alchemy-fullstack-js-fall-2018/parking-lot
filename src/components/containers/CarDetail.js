import { connect } from 'react-redux';
import CarDetail from '../pages/Cars/CarDetail.jsx';
import { getCar } from '../../state/selectors/cars';
import { carArrived, carDeparted } from '../../state/actions/cars';

const mapStateToProps = (state, props) => ({
  car: getCar(state, props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  carDeparted: id => dispatch(carDeparted(id)),
  carArrived: id => dispatch(carArrived(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarDetail);
