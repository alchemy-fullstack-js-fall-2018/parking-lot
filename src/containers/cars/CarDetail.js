import { connect } from 'react-redux';
import CarDetail from '../../components/CarDetail/CarDetail';
import { getCar } from '../../selectors/cars';
import { lotCarArrived, lotCarDeparted } from '../../actions/lot';

const mapStateToProps = (state, props) => {
  return {
    car: getCar(state, props.match.params.id)
  };
};

const mapDispatchToProps = dispatch => ({
  carArrives: (plate) => dispatch(lotCarArrived(plate)),
  carDeparts: (plate) => dispatch(lotCarDeparted(plate))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CarDetail);
