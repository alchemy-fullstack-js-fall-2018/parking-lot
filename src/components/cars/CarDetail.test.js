import React from 'react';
import { shallow } from 'enzyme';
import CarDetail from './CarDetail';
jest.mock('../../routes/routes');

describe('CarDetail component', () => {
  const car = { make: 'Toyota', model: 'Corolla', plate: 'ABC123' };
  const addToLot = jest.fn();
  const removeFromLot = jest.fn();

  it('renders a CarDetail component', () => {
    const wrapper = shallow(<CarDetail
      car={car} inLot={true}
      lotUsage={3}
      addToLot={addToLot}
      removeFromLot={removeFromLot}/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('shows make, model, and plate of car, as well as "yes" if car is inLot and number of times in lot', () => {
    const wrapper = shallow(
      <CarDetail
        car={car}
        inLot={true}
        lotUsage={3}
        addToLot={addToLot}
        removeFromLot={removeFromLot}
      />
    );
    expect(wrapper.html()).toContain('Toyota');
    expect(wrapper.html()).toContain('Corolla');
    expect(wrapper.html()).toContain('ABC123');
    expect(wrapper.html()).toContain('Yes');
    expect(wrapper.html()).toContain(3);
  });

  it('shows "no" if car is not inLot', () => {
    const wrapper = shallow(
      <CarDetail
        car={car}
        inLot={false}
        lotUsage={3}
        addToLot={addToLot}
        removeFromLot={removeFromLot}
      />
    );
    expect(wrapper.html()).toContain('No');
  });

  it('shows "Remove from Lot" on button if car is already in the lot', () => {
    const wrapper = shallow(
      <CarDetail
        car={car}
        inLot={true}
        lotUsage={3}
        addToLot={addToLot}
        removeFromLot={removeFromLot}
      />
    );
    expect(wrapper.find('button').props().children).toContain('Remove from Lot');
  });

  it('shows "Add to Lot" on button if car is already in the lot', () => {
    const wrapper = shallow(
      <CarDetail
        car={car}
        inLot={false}
        lotUsage={3}
        addToLot={addToLot}
        removeFromLot={removeFromLot}
      />
    );
    expect(wrapper.find('button').props().children).toContain('Add to Lot');
  });
});
