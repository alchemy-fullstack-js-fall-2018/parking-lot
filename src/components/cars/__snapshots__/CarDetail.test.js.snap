// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarDetail component renders a CarDetail component 1`] = `
<Fragment>
  <table
    width="100%"
  >
    <tbody>
      <tr>
        <th>
          Make
        </th>
        <th>
          Model
        </th>
        <th>
          Plate
        </th>
        <th>
          In Lot Now
        </th>
        <th>
          Times Used
        </th>
      </tr>
      <tr>
        <th>
          Toyota
        </th>
        <th>
          Corolla
        </th>
        <th>
          ABC123
        </th>
        <th>
          Yes
        </th>
        <th>
          3
        </th>
      </tr>
    </tbody>
  </table>
  <button
    onClick={[Function]}
  >
    Remove from Lot
  </button>
</Fragment>
`;
